src/models/fhir/parse.ts

```diff
@@ -0,0 +1,167 @@
+import { R4 } from "@ahryman40k/ts-fhir-types";
+import {
+  Bundle,
+  Patient,
+  Specimen,
+  Observation,
+  Practitioner,
+  Organization,
+  Device,
+} from "./types";
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { Fhir } = require("fhir");
+
+const fhir = new Fhir();
+
+/**
+ * The mapping definition of each FHIR resource.
+ *
+ * Note: This function will ONLY validate a resource against FHIR base spec.
+ * @param resource Raw FHIR resource
+ * @returns Parsed FHIR resource (simplified)
+ */
+const parsers = (resource: R4.IResourceList | undefined) => {
+  if (!resource) return undefined; // Skip parsing an undefined resource
+
+  const validator = fhir.validate(resource, { errorOnUnexpected: true });
```

Comment:
considering you run this on the bundle, doesn't it mean the bundled resources are verified twice?

- I would split validation from parsing/transformation

Reply:
The `fhir.validate()` here only helps to check against the base spec of FHIR to ensure the structure, cardinality, values, bindings, etc of the a FHIR resource. I feel it's only right to validate a resource first before parsing. If not, it will just parse and miss out some values without knowing.

Another alternative is to get the caller of this helper function to manually call `validate()` first but not sure how to enforce it.

Reply:
but if you validate the bundle, every resource of the bundle will be validated?

Reply:
Yes, it will traverse every resource in the bundle as well. For example, I provided an invalid `resourceType` for the last entry:

```javascript
// Validation against FHIR base spec has failed for Bundle:
[
  {
    "location": "Specimen.type",
    "resourceId": "Bundle.entry[1].resource",
    "severity": "info",
    "message": "Value set \"http://terminology.hl7.org/ValueSet/v2-0487\" could not be found."
  },
  {
    "location": "Observation.category[0]",
    "resourceId": "Bundle.entry[2].resource",
    "severity": "warning",
    "message": "Code \"840539006\" (http://snomed.info/sct) not found in value set"
  },
  // ...
  {
    "location": "",
    "resourceId": "Bundle.entry[6].resource",
    "severity": "fatal",
    "message": "Resource does not have resourceType property, or value is not a valid resource type."
  }
]
```